### PR TITLE
Bias, bug fixes in matching, homo shell 

### PR DESCRIPTION
**Pull request by VasilievaKA**

### CODE REVIEW

File: hypex/operators/operators.py [hunk 1]
```diff
@@ -59,9 +64,29 @@ def execute(self, data: ExperimentData) -> ExperimentData:
             group_field=group_field,
             target_fields=target_fields,
             metric=self.metric,
+            bias_estimation=self.bias_estimation,
         )
         return self._set_value(data, compare_result)
 
+    @staticmethod
+    def bias_coefs(matches, Y_m, X_m):
```
Comment (TonyKatkov89) [inline hypex/operators/operators.py:72]: types not defined

File: hypex/operators/operators.py [hunk 2]
```diff
@@ -59,9 +64,29 @@ def execute(self, data: ExperimentData) -> ExperimentData:
...
+        Y = Y_m[flat_idx]
+        X = np.empty((N, K + 1))
+        X[:, 0] = 1  # intercept term
+        X[:, 1:] = X_m[flat_idx]
+
+        return np.linalg.lstsq(X, Y)[0][1:]  # don't need intercept coef
+
+    @staticmethod
+    def bias(X, X_m, matches, coefs):
+        X_m_mean = [X_m[idx].mean(0) for idx in matches]
```
Comment (TonyKatkov89) [inline hypex/operators/operators.py:85]: Is it possible to just do X_m[matches].mean(0) and to convert to list?

File: hypex/operators/operators.py [hunk 3]
```diff
@@ -59,9 +64,29 @@ def execute(self, data: ExperimentData) -> ExperimentData:
...
+
+        Y = Y_m[flat_idx]
+        X = np.empty((N, K + 1))
+        X[:, 0] = 1  # intercept term
+        X[:, 1:] = X_m[flat_idx]
+
+        return np.linalg.lstsq(X, Y)[0][1:]  # don't need intercept coef
+
+    @staticmethod
+    def bias(X, X_m, matches, coefs):
```
Comment (TonyKatkov89) [inline hypex/operators/operators.py:84]: no types defined

**Review by TonyKatkov89 — APPROVED** [review]
(❁´◡`❁)

File: hypex/operators/operators.py [hunk 4]
```diff
@@ -75,39 +101,172 @@ def _execute_inner_function(
...
-    ) -> Any:
-        raise NotImplementedError
+        cls,
+        data: Dataset,
+        test_data: Optional[Dataset] = None,
+        target_fields: Optional[List[str]] = None,
+        features_fields: Optional[List[str]] = None,
+        **kwargs
+    ) -> Dict:
+        if target_fields is None or features_fields is None or test_data is None:
```
Comment (TonyKatkov89) [inline hypex/operators/operators.py:235]: Why do we have these variables as optional arguments if we raise error if they are None?

**Review by TonyKatkov89 — CHANGES_REQUESTED** [review]

**Review by TonyKatkov89 — APPROVED** [review]
(^///^)

File: hypex/dataset/dataset.py [hunk 5]
```diff
@@ -549,9 +551,11 @@ def dot(self, other: "Dataset") -> "Dataset":
 
     def transpose(
         self,
-        roles: Optional[Union[Dict[Union[str, int], ABCRole], List]] = None,
+        roles: Optional[Union[Dict[str, ABCRole], List]] = None,
     ) -> "Dataset":
-        roles_names = roles.keys() or {} if isinstance(roles, Dict) else roles
+        roles_names: List[str] = (
```
[the remark this answers is not in the document]
Comment (VasilievaKA) [reply hypex/dataset/dataset.py:557]: ?

File: hypex/dataset/dataset.py [hunk 6]
```diff
@@ -252,9 +253,8 @@ def create_empty(backend=BackendsEnum.pandas, roles=None, index=None) -> "Datase
     def _convert_data_after_agg(self, result) -> Union["Dataset", float]:
         if isinstance(result, float):
             return result
-        return Dataset(
-            data=result, roles={column: StatisticRole() for column in self.roles}
-        )
+        role: ABCRole = StatisticRole()
```
Comment (TonyKatkov89) [inline hypex/dataset/dataset.py:256]: Didn't get wgy do we define "role" here and in the next line use it as an iterator.

**Review by TonyKatkov89 — APPROVED** [review]
༼ つ ◕_◕ ༽つ

File: hypex/dataset/dataset.py (hunk 6 above)
Comment (VasilievaKA) [reply hypex/dataset/dataset.py:256]: It is for mypy fixes, we need type initialization before usage

**Review by Dmatryus — APPROVED** [review]
Minor fixes for clean code by comments

**Pull request merged**